src/tools/rbd/action/Bench.cc

```diff
@@ -336,44 +370,40 @@ int do_bench(librbd::Image& image, io_type_t io_type,
       ++cur_ios;
       cur_off += io_size;
 
-      if (read_flag)
+      if (read_flag) {
         read_ops++;
-      else
+      } else {
         write_ops++;
-    }
-
-    if (r < 0) {
-      break;
-    }
+      }
 
-    // Set the thread_offsets of next I/O
-    for (i = 0; i < io_threads; ++i) {
       switch (io_pattern) {
       case IO_PATTERN_RAND:
-        thread_offset[i] = (rand() % (size / io_size)) * io_size;
-        continue;
+        pos = (io_start % io_size) + (rand() % num_chunks) * io_size;
+        break;
       case IO_PATTERN_SEQ:
-        if (off < (seq_chunk_length * io_threads)) {
-          thread_offset[i] += io_size;
-        } else {
-          // thread_offset is adjusted to the chunks unassigned to threads.
-          thread_offset[i] = off + (i * io_size);
+	// Advance by chunk length to get pos for next thread
+        pos += seq_chunk_length;
+        if (i < rem_chunks) {
+          pos += io_size;
         }
-        if (thread_offset[i] + io_size > size) {
-          thread_offset[i] = seq_chunk_length * i;
```

Comment:
Note bug in old code - if io_bytes is greater than image size then after sequentially accessing whole image once the threads get stuck repeatedly accessing this offset because off > size